common/kvstore/src/main/java/org/apache/spark/util/kvstore/InMemoryStore.java

```diff
@@ -322,15 +370,28 @@ private static Object indexValueForEntity(KVTypeInfo.Accessor getter, Object ent
      */
     private List<T> copyElements() {
       if (parent != null) {
-        KVTypeInfo.Accessor parentGetter = ti.getParentAccessor(index);
-        Preconditions.checkArgument(parentGetter != null, "Parent filter for non-child index.");
-        Comparable<?> parentKey = asKey(parent);
-
-        return elements.stream()
-          .filter(e -> compare(e, parentGetter, parentKey) == 0)
-          .collect(Collectors.toList());
+        Comparable<Object> parentKey = asKey(parent);
+        if (!naturalParentIndexName.isEmpty() &&
+          naturalParentIndexName.equals(ti.getParentIndexName(index))) {
+          NaturalKeys children =
+            parentToChildrenMap.computeIfAbsent(parentKey, k -> new NaturalKeys());
+          ArrayList<T> elements = new ArrayList<>();
+          for (Comparable<Object> naturalKey : children.keySet()) {
+            data.computeIfPresent(naturalKey, (k, v) -> {
```

Comment:
is this faster than `data.values().stream()...`?

Reply:
That's a linear scan, which we are trying to avoid.

Reply:
ditto, let's highlight why we need to do linear scan in the else branch.